ircd/privilege.c

```diff
@@ -58,6 +66,178 @@ privilegeset_get_any(const char *name)
 	return NULL;
 }
 
+static int
+privilegeset_cmp_priv(const void *a_, const void *b_)
+{
+	const char *const *a = a_, *const *b = b_;
+	return strcmp(*a, *b);
+}
+
+static void
+privilegeset_index(struct PrivilegeSet *set)
+{
+	size_t n;
+	const char *s;
+	const char **p;
+
+	if (set->privs != NULL)
+		rb_free(set->privs);
+
+	set->privs = rb_malloc(sizeof *set->privs * (set->size + 1));
+	p = set->privs;
+
+	for (n = 0, s = set->priv_storage; n < set->size; n++, s += strlen(s) + 1)
+		*p++ = s;
+	qsort(set->privs, set->size, sizeof *set->privs, privilegeset_cmp_priv);
+	set->privs[set->size] = NULL;
+}
+
+static void
+privilegeset_add_privs(struct PrivilegeSet *dst, const char *privs)
+{
+	size_t cur_size, alloc_size;
+	size_t n;
+
+	if (dst->priv_storage == NULL)
+	{
+		dst->stored_size = dst->allocated_size = 0;
+		cur_size = 0;
+		alloc_size = 256;
+	}
+	else
+	{
+		cur_size = dst->stored_size;
+		alloc_size = dst->allocated_size;
+	}
+
+	dst->stored_size = cur_size + strlen(privs) + 1;
+
+	while (alloc_size < dst->stored_size)
+		alloc_size *= 2;
+
+	if (alloc_size > dst->allocated_size)
+		dst->priv_storage = rb_realloc(dst->priv_storage, alloc_size);
+
+	dst->allocated_size = alloc_size;
+
+	const char *s;
+	char *d;
+	for (s = privs, d = dst->priv_storage; s < privs + strlen(privs); s += n , d += n)
+	{
+		const char *e = strchr(s, ' ');
+		/* up to space if there is one, else up to end of string */
+		n = 1 + (e != NULL ? e - s : strlen(s));
+		rb_strlcpy(d, s, n);
+
+		dst->size += 1;
+	}
+
+	privilegeset_index(dst);
+}
+
+static void
+privilegeset_add_privilegeset(struct PrivilegeSet *dst, const struct PrivilegeSet *src)
+{
+	size_t cur_size, alloc_size;
+
+	if (dst->priv_storage == NULL)
+	{
+		dst->stored_size = dst->allocated_size = 0;
+		cur_size = 0;
+		alloc_size = 256;
+	}
+	else
+	{
+		cur_size = dst->stored_size;
+		alloc_size = dst->allocated_size;
+	}
+
+	dst->stored_size = cur_size + src->stored_size;
+
+	while (alloc_size < dst->stored_size)
+		alloc_size *= 2;
+
+	if (alloc_size > dst->allocated_size)
+		dst->priv_storage = rb_realloc(dst->priv_storage, alloc_size);
+
+	dst->allocated_size = alloc_size;
+
+	memcpy(dst->priv_storage + cur_size, src->priv_storage, src->stored_size);
+	dst->size += src->size;
+
+	privilegeset_index(dst);
+}
+
+static struct PrivilegeSet *
+privilegeset_new_orphan(const char *name)
+{
+	struct PrivilegeSet *set;
+	set = rb_malloc(sizeof *set);
+	set->size = 0;
+	set->privs = NULL;
+	set->priv_storage = NULL;
+	set->status = 0;
+	set->refs = 0;
+	set->name = rb_strdup(name);
+	return set;
+}
+
+static void
+privilegeset_free(struct PrivilegeSet *set)
```

Comment:
it's nice when *free* functions tolerate `NULL`. Instead of testing `set -> shadow` for `NULL`, let's test `set` for `NULL` and then we can unconditionally free `->shadow` like we unconditionally free everything else.